### PR TITLE
feat(cli): support reserved environment variable "APISIX_DEPLOYMENT_E…

### PR DESCRIPTION
**Pull request by An-DJ**
…TCD_HOST"

### Description

The user can use env variable "APISIX_DEPLOYMENT_ETCD_HOST" to config the etcd.host.

Priority: env var > config.yaml

### Checklist

- [X] I have explained the need for this PR and the problem it solves
- [X] I have explained the changes or the new features added to this PR
- [X] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [X] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: t/cli/test_main.sh [hunk 1]
```diff
@@ -211,6 +211,17 @@ fi
 
 echo "passed: resolve variables"
 
+# support reserved environment variable APISIX_DEPLOYMENT_ETCD_HOST
+
+echo '' > conf/config.yaml
```
Comment (spacewander) [inline t/cli/test_main.sh:216]: Let's add a config to check if the env can override config

File: t/cli/test_main.sh [hunk 2]
```diff
@@ -211,6 +211,17 @@ fi
...
+out=$(APISIX_DEPLOYMENT_ETCD_HOST='["http://127.0.0.1:2333"]' make init 2>&1 || true)
```
Comment (spacewander) [inline t/cli/test_main.sh:217]: Have you tested it with ./bin/apisix start?

File: apisix/cli/file.lua [hunk 3]
```diff
@@ -115,6 +117,21 @@ end
 _M.resolve_conf_var = resolve_conf_var
 
 
+local function replace_by_reserved_env_vars(conf)
+    -- TODO: support more reserved environment variables
+    local v = getenv("APISIX_DEPLOYMENT_ETCD_HOST")
```
Comment (tokers) [inline apisix/cli/file.lua:122]: Since it's an array, let's call it `APISIX_DEPLOYMENT_ETCD_HOSTS`.

Comment (tokers) [reply apisix/cli/file.lua:122]: 👍🏻 Good point.

File: t/cli/test_main.sh [hunk 4]
```diff
@@ -211,6 +211,32 @@ fi
...
+      - "http://127.0.0.1:2379"
+' > conf/config.yaml
+
+out=$(APISIX_DEPLOYMENT_ETCD_HOST='["http://127.0.0.1:2333"]' make init 2>&1 || true)
+if ! echo "$out" | grep "connection refused" > /dev/null; then
+    echo "failed: failed to configure etcd host with reserved environment variable"
+    exit 1
+fi
+
+out=$(APISIX_DEPLOYMENT_ETCD_HOST='["http://127.0.0.1:2333"]' make run 2>&1 || true)
```
Comment (spacewander) [inline t/cli/test_main.sh:232]: Can you configure a connectable address and check if it works?

Comment (An-DJ) [reply t/cli/test_main.sh:242]: Done.

**Review by spacewander — COMMENTED** [review]
It seems this feature is broken, when I tried "valid host in env" + "invalid host in conf" + "make run".

**Comment by An-DJ** [comment]
> It seems this feature is broken, when I tried "valid host in env" + "invalid host in conf" + "make run".

@spacewander This case is covered by the latest commit and passed.
https://github.com/apache/apisix/pull/8898/files#diff-cb891f9ef4943580aa92827c7d725f3c717923cb85344cc6d8195d70f7ef01b0R236

**Comment by spacewander** [comment]
> > It seems this feature is broken, when I tried "valid host in env" + "invalid host in conf" + "make run".
> 
> @spacewander This case is covered by the latest commit and passed. [#8898 (files)](https://github.com/apache/apisix/pull/8898/files#diff-cb891f9ef4943580aa92827c7d725f3c717923cb85344cc6d8195d70f7ef01b0R236)

I spend some time playing with this feature locally. Look like when APISIX is running, it will read the env variable in a different path. The CI stops APISIX too early, so APISIX doesn't have time to read it.

File: apisix/cli/ngx_tpl.lua [hunk 5]
```diff
@@ -51,6 +51,9 @@ worker_shutdown_timeout {* worker_shutdown_timeout *};
 env APISIX_PROFILE;
 env PATH; # for searching external plugin runner's binary
 
+# reserved environment variables for configuration
+env APISIX_DEPLOYMENT_ETCD_HOST;
```
Comment (spacewander) [inline apisix/cli/ngx_tpl.lua:55]: Can we find a way to cover it in the CI? So we don't need to spend time verifying it locally next time.

File: t/cli/test_main.sh [hunk 6]
```diff
@@ -211,6 +211,42 @@ fi
...
+
+out=$(APISIX_DEPLOYMENT_ETCD_HOST='["http://127.0.0.1:2379"]' make init 2>&1 || true)
+if echo "$out" | grep "connection refused" > /dev/null; then
+    echo $failed_msg
+    exit 1
+fi
+
+out=$(APISIX_DEPLOYMENT_ETCD_HOST='["http://127.0.0.1:2379"]' make run 2>&1 || true)
+if echo "$out" | grep "connection refused" > /dev/null; then
+    echo $failed_msg
```
Comment (spacewander) [inline t/cli/test_main.sh:236]: We should make stop here

File: t/cli/test_main.sh [hunk 7]
```diff
@@ -211,6 +211,42 @@ fi
...
+    exit 1
+fi
+
+if ! grep "env APISIX_DEPLOYMENT_ETCD_HOST;" conf/nginx.conf > /dev/null; then
+    echo "failed: 'env APISIX_DEPLOYMENT_ETCD_HOST;' not in nginx.conf"
+    echo $failed_msg
```
Comment (spacewander) [inline t/cli/test_main.sh:242]: Ditto

**Review by spacewander — APPROVED** [review]

**Review by soulbird — APPROVED** [review]

**Pull request merged**